fitbenchmarking/cli/checkpoint_handler.py

```diff
@@ -110,6 +136,205 @@ def generate_report(options_file='', additional_options=None, debug=False):
+    """
+    Combine multiple checkpoint files into one following these rules:
+     1) The output will be the same as combining them one at a time in
+        sequence. i.e. A + B + C = (A + B) + C
+        The rules from here on will only reference A and B as the relation is
+        recursive.
+     2) Datasets
+         2a) Datasets in A and B are identical if they agree on the label
+         2b) If A and B contain identical datasets, the problems and results
+             are combined as below.
+             The remainder of these rules assume that A and B are identical
+             datasets as the alternative is trivial.
+     3) Problems
+         3a) If problems in A and B agree on name, ini_params, ini_y, x, y,
+             and e then these problems are considered identical.
+         3b) If A and B share identical problems, the details not specified in
+             3a are taken from A.
+         3c) If problems in A and B are not identical but share a name, the
+             name of the project in B should be updated to "<problem_name>*".
+     4) Results
+         4a) If results in A and B have identical problems and agree on name,
+             software_tag, minimizer_tag, jacobian_tag, hessian_tag, and
+             costfun_tag they are considered identical.
+         4b) If A an B share identical results, the details not specified in 4a
+             are taken from A if strategy is 'first', or B if strategy is
+             'last'.
+     5) As tables are grids of results, combining arbitrary results can lead to
+        un-table-able checkpoint files.
+        This occurs when the problems in A and B are not all identical and the
+        set of combinations of software_tag, minimizer_tag, jacobian_tag,
+        hessian_tag, and costfun_tag for which there are results in each of A
+        and B are not identical.
+        E.g. B has a problem not in A and uses a minimizer for which there are
+             no results in A.
+         5a) If the resulting checkpoint file would have the above issue, the
+             checkpoints are incompatible.
+         5b) Incompatible checkpoint files can be combined but should raise
+             warnings and mark the dataset in the checkpoint file.
+             Note: Some datasets may be incompatible where others can be
+                   successfully combined.
+     6) Unselected minimizers and failed problems will be discarded when
+        combining.
+
+    :param files: The files to combine.
+    :type files: list[str]
+    :param output: The name for the new checkpoint file.
+    :type output: str
+    :param strategy: The stategy for merging identical elements.
+                     Options are 'first' or 'last'.
+    :type strategy: str
+    :param debug: Enable debugging output.
+    :type debug: bool
+    """
+    if len(files) < 2:
+        return
+    if strategy == 'first':
+        pass
+    elif strategy == 'last':
+        files = list(reversed(files))
```

Comment:
This is neater. I'm a bit wary of changing an input inplace (could lead to side effects outside of the function), as a simple example of what I'm concerned about:
```
def print_reversed(l):
    l.reverse()
    print(l)

a = [1, 2, 3]
print_reversed(a)
# [3, 2, 1]
print_reversed(a)
# [1, 2, 3]
```

But I don't think this is going to be an issue with the way the function is used at the moment.

Reply:
It is valid point. I think you are right that changing a variable in place can lead to issues. But do you think we would be fine because it is a `parser`? it will run the merge function with the arguments and discard them.

Reply:
When testing I spotted I already had logic for the accuracy strat so I reworked it to be a bit neater (I think). This bit is now not a problem and we don't have to iterate over the list the extra time to flip it.